repos/system_upgrade/el7toel8/libraries/pam.py

```diff
@@ -0,0 +1,54 @@
+import re
+
+
+class PAM(object):
+    files = [
+        '/etc/pam.d/system-auth',
+        '/etc/pam.d/smartcard-auth',
+        '/etc/pam.d/password-auth',
+        '/etc/pam.d/fingerprint-auth',
+        '/etc/pam.d/postlogin'
+    ]
+
+    def __init__(self, config):
+        self.modules = []
+        self.parse(config)
+
+    def parse(self, config):
+        result = re.findall(
+            r"^[ \t]*[^#\s]+.*(pam_\S+)\.so.*$",
+            config,
+            re.MULTILINE
+        )
+
+        self.modules = result
+
+    def has(self, module):
+        return module in self.modules
+
+    def has_unknown_module(self, known_modules):
+        for module in self.modules:
+            if module not in known_modules:
+                print(module)
+                return True
+
+        return False
+    
+    @staticmethod
+    def read_file(file):
```

Comment:
Not sure you really need to handle those exceptions, see comment above about read_file